[AArch64][GlobalISel] Legalize ptr shuffle vector to s64

davemgreen opened the pull request:
This converts all ptr element shuffle vectors to s64, so that the existing vector legalization handling can lower them as needed. This prevents a lot of fallbacks that currently try to generate things like `<2 x ptr> G_EXT`.

I'm not sure if bitcast/inttoptr/ptrtoint is intended to be necessary for vectors of pointers, but it uses buildCast for the casts, which now generates a ptrtoint/inttoptr.

madhur13490 commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1077,7 +1079,16 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
       .clampNumElements(0, v4s32, v4s32)
-      .clampNumElements(0, v2s64, v2s64);
+      .clampNumElements(0, v2s64, v2s64)
+      .bitcastIf(
```
Instead of doing bitcast, why not make v2p0 legal? Generating bitcast and going through PTRTOINT seems sub-optimal.

davemgreen replied:
Three are two options we have, both of which should produce the same output assembly. Either we try and keep v2p0 shuffles as legal, which means having to handle them in every shuffle combine that we add and legalized all the various vNp0 types correctly, which are easy to miss. The alternative like this patch does it to convert them to v2s64 and just deal with that single type.

Because of the relative rarity of pointer shuffles it feels fine to convert them to s64 and let that legalizing happen as a s64 operation.

madhur13490 replied:
Alright, thanks for the explanation.

arsenm commented on llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:
```diff
@@ -1077,7 +1079,16 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
+      .clampNumElements(0, v2s64, v2s64)
+      .bitcastIf(
+          // Bitcast pointers vector to i64.
+          [=](const LegalityQuery &Query) {
+            return Query.Types[0].isPointerVector();
+          },
```
Should probably have this in the prebuilt set of predicates along with isPointer and isVector

arsenm commented on llvm/test/CodeGen/AArch64/shufflevector.ll:
```diff
@@ -393,12 +388,18 @@ define <4 x i64> @shufflevector_v4i64(<4 x i64> %a, <4 x i64> %b) {
+; CHECK-GI-NEXT:    ret
     %c = shufflevector <4 x ptr> %a, <4 x ptr> %b, <4 x i32> <i32 1, i32 3, i32 5, i32 7>
     ret <4 x ptr> %c
 }
```
Check a 3 x ptr case?

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -3697,6 +3697,41 @@ LegalizerHelper::bitcastConcatVector(MachineInstr &MI, unsigned TypeIdx,
+LegalizerHelper::LegalizeResult
+LegalizerHelper::bitcastShuffleVector(MachineInstr &MI, unsigned TypeIdx,
+                                      LLT CastTy) {
+  auto ShuffleMI = dyn_cast<GShuffleVector>(&MI);
```
Use `cast<>`

arsenm reviewed (commented):
This is fine, but it would be better to directly consume these with the same legality as the equivalent integer operation

davemgreen committed:
[AArch64][GlobalISel] Legalize ptr shuffle vector to s64

This converts all ptr element shuffle vectors to s64, so that the existing
vector legalization handling can lower them as needed.

I'm not sure if bitcast/inttoptr/ptrtoint is intended to be necessary for
vectors of pointers, but it uses buildCast for the casts, which now generates a
ptrtoint/inttoptr.

davemgreen commented:
> This is fine, but it would be better to directly consume these with the same legality as the equivalent integer operation

We might change how this works in the future, but currently quite a lot is falling back. As we figure out how different float types will work it might be worth considering vectors of pointers at the same time. Ideally they would not work very differently to a s64 and we would not need to handle them both separately for the operations that only depend upon the size.
